### PR TITLE
[Core][Metrics] Fix metrics counter treated as gauge

### LINKED ISSUE #37768: [Metrics] Application metrics API Counter create Gauge when imported to Prometheus
### What happened + What you expected to happen

When running the below script, the counter application metrics are imported as gauge. This should be fixed because it doesn't conform the prometheus standard.

### Versions / Dependencies

master

### Reproduction script

```
from ray.util.metrics import Gauge, Histogram, Counter
RAY_METRIC_PREFIX = "mp" 
DEFAULT_METRIC_BOUNDS = [0.1,1]
histogram_test = Histogram(
    f"{RAY_METRIC_PREFIX}_h_test",
    description="some value",
    boundaries=DEFAULT_METRIC_BOUNDS,
    tag_keys=("algo",),
)
histogram_test.observe(0.5, {"algo": "magic"})

gauge_test = Gauge(
    f"{RAY_METRIC_PREFIX}_g_test",
    description="some value",
    tag_keys=("algo",),
)
gauge_test.set(4, {"algo": "magic"})

counter_test = Counter(
    f"{RAY_METRIC_PREFIX}_c_test",
    description="some value",
    tag_keys=("algo",),
)
counter_test.inc(1, {"algo": "magic"})
```

### Issue Severity

None

### PR DESCRIPTION
**Pull request by jonathan-anyscale**
## Why are these changes needed?
Currently in `metric_agent.py`, any metric with `double` value is converted to `LastValueAggregationData` which then converted to prometheus Gauge stats. The issue is `Counter` from `ray.utils.metrics` used `double` value as well which results it's being treated as Gauge in the prometheus by `OpencensusProxyMetric::record`

context: we want to fix `metric.Counter` that appeared as `Gauge` in prometheus. The problem can be solved by switching `GaugeMetricFamily`  to `CounterMetricFamily` when converting opencensus format to prometheus format. However, doing this changes the metric name by appending `_total` and since ray dashboard, ray serve uses `metric.Counter` a lot, this can cause massive breaking change.
The actual problem is _total  is added inside [prometheus_client](https://github.com/prometheus/client_python/blob/2dcd17efd0ce2f0a1ad15cb3c150ffcdc42ced65/prometheus_client/exposition.py#L257) in the `generate_latest` which is called by internal function `_bake_output` so there's no way to modify this outside the prometheus client itself.

Solution: vendoring prometheus client + modify it to prevent the breaking change

## Related issue number
Closes #37768 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/metrics_agent.py [hunk 1]
```diff
@@ -371,6 +379,18 @@ def to_metric(
             )
             return metric
 
+        elif isinstance(agg_data, SumAggregationData):
```
Comment (rkooo567) [inline python/ray/_private/metrics_agent.py:382]: Can you add a comment why we convert Sum -> Counter (basically opencensus sum allows to increment float > 1)

**Review by rkooo567 — COMMENTED** [review]
nice! can you add a unit test that checks

1. export histogram, gauge, and counter
2. check prom endpoint
3. verify all types are correct

you can see `test_metrics_agent.py` as a reference

File: python/ray/_private/metrics_agent.py [hunk 2]
```diff
@@ -371,6 +379,18 @@ def to_metric(
...
+            metric = metrics_map.get(metric_name)
+            if not metric:
+                metric = CounterMetricFamily(
```
Comment (jonathan-anyscale) [inline python/ray/_private/metrics_agent.py:385]: So apparently GaugeMetricFamily (gauge), CounterMetricFamily (Count) and UnknownMetricFamily (Sum) from prometheus is modifying the metrics name...
For example `internal_num_processes_skipped_job_mismatch` which initially in `GaugeMetricFamily` since it's a type Sum becomes `internal_num_processes_skipped_job_mismatch_workers_total` when it's type is changed to `Count` corresponds to CounterMetricFamily

is it okay to change the metrics name resulted from this? Otherwise `ray.utils.metrics.Counter` and any `Sum` stats like `internal_num_processes_skipped_job_mismatch` will be treated as untyped

**Review by architkulkarni — APPROVED** [review]
Approving changes in `python/ray/autoscaler` as codeowner

File: ci/docker/base.test.Dockerfile [hunk 3]
```diff
@@ -60,3 +60,4 @@ WORKDIR /ray
 COPY . .
 
 RUN ./ci/env/install-bazel.sh
+
```
Comment (aslonnie) [inline ci/docker/base.test.Dockerfile:63]: why? maybe just revert this change? (and I do not need to review this PR.)

Comment (jonathan-anyscale) [reply ci/docker/base.test.Dockerfile:63]: it was for some cache-related issue test, will remove!!
cc: @can-anyscale

**Review by rkooo567 — COMMENTED** [review]
can you tell me what file within prometheus client you changed?

**Comment by jonathan-anyscale** [comment]
> can you tell me what file within prometheus client you changed?

metrics_core.py: Adding new `SumMetricFamily` with type counter 
exposition.py: in `generate_latest`  check if Metric has type counter and is `SumMetricFamily`, don't append the `_total` to metric name

File: python/ray/_private/thirdparty/prometheus_client/__init__.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (jonathan-anyscale) [reply python/ray/_private/thirdparty/prometheus_client/__init__.py:1]: I added the license to ray LICENSE

**Review by rkooo567 — COMMENTED** [review]
A couple more feedback; 

1. Let's use the same way as https://github.com/ray-project/ray/pull/41426#issuecomment-1834620507 (still seems to be in progress)
2. Maybe consider patch instead of directly modifying the source so that we can track changes

**Comment by rkooo567** [comment]
The priority of the PR is not clear yet (whether or not it is a blocker). Let's wait one more day for the feedback

**Review by edoakes — APPROVED** [review]
Stamp for requirements changes/vendoring.

File: ci/env/check_minimal_install.py [hunk 5]
```diff
@@ -20,7 +20,6 @@
     "py-spy",
     "gpustat",
     "opencensus",
-    "prometheus_client",
```
Comment (can-anyscale) [inline ci/env/check_minimal_install.py:23]: why is this removed?

Comment (can-anyscale) [reply ci/env/check_minimal_install.py:23]: ah got it now

**Comment by jjyao** [comment]
Replaced by #43795
